File: src/main/java/com/sourcegraph/cody/CodyFileEditorListener.java
package com.sourcegraph.cody;

import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.fileEditor.FileDocumentManager;
import com.intellij.openapi.fileEditor.FileEditorManager;
import com.intellij.openapi.fileEditor.FileEditorManagerListener;
import com.intellij.openapi.vfs.VirtualFile;
import com.sourcegraph.cody.agent.CodyAgent;
import com.sourcegraph.cody.agent.CodyAgentCodebase;
import com.sourcegraph.cody.agent.CodyAgentService;
import com.sourcegraph.cody.agent.protocol.ProtocolTextDocument;
import org.jetbrains.annotations.NotNull;

public class CodyFileEditorListener implements FileEditorManagerListener {
  @Override
  public void fileOpened(@NotNull FileEditorManager source, @NotNull VirtualFile file) {
    CodyAgentService.withAgent(source.getProject(), agent -> fileOpened(source, agent, file));
  }

  @Override
  public void fileClosed(@NotNull FileEditorManager source, @NotNull VirtualFile file) {
    var protocolTextFile = ProtocolTextDocument.fromVirtualFile(source, file);
    CodyAgentService.withAgent(
        source.getProject(), agent -> agent.getServer().textDocumentDidClose(protocolTextFile));
  }

  public static void fileOpened(
      @NotNull FileEditorManager source, CodyAgent codyAgent, @NotNull VirtualFile file) {
    var project = source.getProject();
    ApplicationManager.getApplication()
        .invokeLater(
            () -> {
              var textDocument = ProtocolTextDocument.fromVirtualFile(source, file);
              codyAgent.getServer().textDocumentDidOpen(textDocument);
              CodyAgentCodebase.getInstance(project).onFileOpened(file);
            });
  }

  public static void editorChanged(@NotNull Editor editor) {
    var project = editor.getProject();
    var file = FileDocumentManager.getInstance().getFile(editor.getDocument());
    if (project != null && file != null) {
      var fileEditorManager = FileEditorManager.getInstance(project);
      CodyAgentService.withAgent(project, agent -> fileOpened(fileEditorManager, agent, file));
    }
  }
}

File: src/main/kotlin/com/sourcegraph/cody/listeners/CodyFileEditorListener.kt
package com.sourcegraph.cody.listeners

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.editor.EditorFactory
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.fileEditor.FileEditorManagerListener
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.sourcegraph.cody.agent.CodyAgent
import com.sourcegraph.cody.agent.CodyAgentService.Companion.withAgent
import com.sourcegraph.cody.agent.protocol.ProtocolTextDocument.Companion.fromVirtualFile

class CodyFileEditorListener : FileEditorManagerListener {
  private val logger = Logger.getInstance(CodyFileEditorListener::class.java)

  override fun fileOpened(source: FileEditorManager, file: VirtualFile) {
    try {
      source.selectedTextEditor?.let { editor ->
        val protocolTextFile = fromVirtualFile(editor, file)
        withAgent(source.project) { agent: CodyAgent ->
          agent.server.textDocumentDidOpen(protocolTextFile)
        }
      }
    } catch (x: Exception) {
      logger.warn("Error in fileOpened method for file: ${file.path}", x)
    }
  }

  override fun fileClosed(source: FileEditorManager, file: VirtualFile) {
    try {
      source.selectedTextEditor?.let { editor ->
        val protocolTextFile = fromVirtualFile(editor, file)
        withAgent(source.project) { agent: CodyAgent ->
          agent.server.textDocumentDidClose(protocolTextFile)
        }
      }
    } catch (x: Exception) {
      logger.warn("Error in fileClosed method for file: ${file.path}", x)
    }
  }

  companion object {
    private val logger = Logger.getInstance(CodyFileEditorListener::class.java)

    fun registerAllOpenedFiles(project: Project, codyAgent: CodyAgent) {

      ApplicationManager.getApplication().invokeLater {
        val fileDocumentManager = FileDocumentManager.getInstance()

        EditorFactory.getInstance().allEditors.forEach { editor ->
          fileDocumentManager.getFile(editor.document)?.let { file ->
            try {
              val textDocument = fromVirtualFile(editor, file)
              codyAgent.server.textDocumentDidOpen(textDocument)
            } catch (x: Exception) {
              logger.warn("Error calling textDocument/didOpen for file: ${file.path}", x)
            }
          }
        }

        if (project.isDisposed) return@invokeLater
        FileEditorManager.getInstance(project).selectedTextEditor?.let { editor ->
          val file = fileDocumentManager.getFile(editor.document)
          try {
            val textDocument = fromVirtualFile(editor, file!!)
            codyAgent.server.textDocumentDidFocus(textDocument)
          } catch (x: Exception) {
            logger.warn("Error calling textDocument/didFocus on ${file?.path}", x)
          }
        }
      }
    }
  }
}
